prime-router/docs/proposals/0013-batch-retry.md

```diff
@@ -0,0 +1,66 @@
+# Batch Process Update / Retry
+The batching process could use more robust retry functionality and there is an opportunity for additional improvements
+to the functionality as a whole to address scalability and performance concerns.
+
+Associated Ticket - #2902
+
+### Current Batch Functionality  
+Each incoming message puts an entry into the 'batch' queue per receiver that message is going to, using the hidden
+'next action at' part of the queue and selecting the next 'batch bucket' for the associated receiver.
+
+When that time comes all queue messages for the receiver in question are added to the queue and picked up by individual 
+instances of the batch function (one per message on the queue). 
+
+Each of these functions then race to select and lock 
+all records from the Task table that are for the receiver in question. The function instances that fail to get any
+records die.
+
+If a batch function fails to succeed it will be lost in the ether and require manual resetting to pick up and batch
+included records.
+
+### Proposed New Functionality
+We can improve this situation and add retry/scalability by adding a cron job and altering the way we use the batch
+queue. Instead of adding each message-per-receiver-per-bucket to the queue at ingestion a cron job could run at
+the minimum granularity for batching, determine which receivers need to be batched at that time, and put a message
+on the queue to be handled immediately by the batchFunction. This is a minimal-change approach that leaves the
+Task table management as it is but severely reduces our performance bottlenecks and race conditions. Once done, the
+single run of the batch function for that receiver will pull either all outstanding records (or up to <limit> if there)
+is one for this receiver and batch them.
+
```

Comment:
1. ideally process would stop computing the next_action_at, allowing it to be lighter weight and the system more recoverable... this is assuming it's never the case that a report should be held for future batching (e.g. don't batch and send report A until at least three days from now)
    - I do think it makes sense to consider, and maybe even implement drafts of, a few of ways of figuring out what batches go out when.
2. What happens in this case currently? The simplest solution sounds like iterating through in the reports in chunks of `limit` within the batch function itself.

Reply:
another possibility for the `scheduler` is to run it once a day, hour, whatever and have it use azure queue's scheduled events kind of like we do now. The benefit of this is we wouldn't have to figure out the minimal granularity of batching, and so wouldn't have to make any updates if that was changed.

Reply:
1.   I'm not a fan of removing next_action_at as produced by the process step - its too useful, its already written, and its accurate. 
1.  Note that the 'tweak' gives you retries for free, no extra coding, no complexity.

2. _What happens currently?_   With one message per report now, there are threads-a-plenty to run multiple chunks.  When the alarm goes off, up to 12 batch function threads start.  If a thread locks the first 1000 rows in Task, then there's a zillion other threads ready to lock the next 1000 and so on.
2.   I worry the 'simplest solution' is single threaded, and won't solve the 'horrid' scenario I outlined.   We have to use the cloud's  parallelism to recover, or you risk falling ever further behind.

FYI Minimal granularity of batching right now is 1 minute - that's baked into the settings objects, so isn't likely to change soon.

Reply:
OK, @whytheplatypus and I had a good convo on this.     We're proposing to stick with the original Cle/Gage idea for part 1) of the above

Reply:
I haven't had time to think about this, but I wonder if we could, as part of this work, correct the `limit` so it nicely limits on number-of-Items rather than the current number-of-reports-batched, which is confusing and not how customers think.